### PR TITLE
Introduces Dart layer

### PR DESCRIPTION
**Pull request by bltavares**
This is a initial implementation of a Dart later.
A [previous implementation] has been deprecated, using a forked version of `dart-mode`.

This layer uses the [dart-mode] linked on the [Dart tools] page.
The layer works with a Flutter installation as well.

Ideally, we would be able to offer a `lsp` backend as an option to use, instead
of using the Dart Analyzer, but I'm still figuring out how to implement and
integrate this.

This layer has [some] [fixes] to make it work, and it is waiting for upstream to
take patches.

- Resolves https://github.com/syl20bnr/spacemacs/issues/8333
- Resolves https://github.com/syl20bnr/spacemacs/issues/8395
- Resolves https://github.com/syl20bnr/spacemacs/issues/10452
- Resolves https://github.com/syl20bnr/spacemacs/issues/11631

---
[previous implementation]: https://github.com/syl20bnr/spacemacs/pull/8406
[dart-mode]: https://github.com/bradyt/dart-mode
[Dart tools]: https://www.dartlang.org/tools
[some]: https://github.com/bradyt/dart-mode/pull/67
[fixes]: https://github.com/bradyt/dart-mode/pull/66

### CODE REVIEW

**Comment by bltavares** [comment]
Hi @averrin, what do you think about this layer? I would appreciate some feedback as you had a previous run at it.

**Comment by averrin** [comment]
You can start dansing around lsp from my config for doom:

```elisp
(def-package! lsp-mode)
(def-package! lsp-ui
  :after lsp-mode
  :hook (lsp-mode . lsp-ui-mode)
)

(def-package! company-lsp
  :after lsp-mode
  :config
    (push 'company-lsp company-backends))

(defun averrin//dart-mode-enable ()
  "Init dart-mode"
  (lsp-define-stdio-client
    lsp-dart-major-mode
    "dart"
    (lambda () default-directory)
    '("~/.pub-cache/bin/dart_language_server"))

  ;; (push 'dart-mode flycheck-global-modes)
  (set (make-local-variable 'company-backends)
      '(company-lsp (company-dabbrev)))
  (flycheck-mode)
  (lsp-dart-major-mode-enable)
  (define-key lsp-ui-mode-map [remap xref-find-definitions] #'lsp-ui-peek-find-definitions)
  (define-key lsp-ui-mode-map [remap xref-find-references] #'lsp-ui-peek-find-references)
  ;; (setq lsp-ui-sideline-code-actions-prefix "💡 ")
)

(def-package! dart-mode
    :after company-lsp
    :hook (dart-mode . averrin//dart-mode-enable)
)
```

It has some problems, eg. with applying quick fixes, but works. And i use 'lite' version of dart-mode with syntax highlighting only.
I prefer lsp-way because it can be used with other languages (eg. i use cquery for c++). And if you make good lsp-based solution (even for spacemacs) it will be awesome for me=)

File: layers/+lang/dart/packages.el [hunk 1]
```diff
@@ -0,0 +1,76 @@
...
+;; This file is not part of GNU Emacs.
+;;
+;;; License: GPLv3
+
+;;; Code:
+
+;; TODO: offer lsp integration
+(defconst dart-packages
+  '(dart-mode
+    flycheck))
```
Comment (Compro-Prasad) [inline layers/+lang/dart/packages.el]: Flycheck is an optional dependency. So, I would suggest you to remove this line.

**Comment by Compro-Prasad** [comment]
Dart newbie here. I just checked the features of this layer and it seems to work perfectly fine. @bltavares Are you planning to use lsp?

**Comment by Compro-Prasad** [comment]
It would be nice to have the layer automically installed when I open a dart file.

File: layers/+lang/dart/packages.el (hunk 1 above)
Comment (bltavares) [reply layers/+lang/dart/packages.el]: I'm not sure how dependencies work, but looking at other layers that have flycheck integration, all of them declares flycheck on the packages list.

I was mostly following the same setup, as the documentation shows as well.

**Comment by bltavares** [comment]
@Compro-Prasad The dart analyzer integration is good enough for me, so I'm not currently planning to move to LSP at this moment. There is still a lot of discussing on how to setup the LSP layers on Spacemacs, so I'll let it settle first.

**Comment by bltavares** [comment]
I've just updated the layer to use the updated version of `dart-mode` as well as integrate better with flycheck.

I'm not sure how it would be possible to offer automatic installation when opening a file tho.

File: layers/+lang/dart/packages.el (hunk 1 above)
Comment (Compro-Prasad) [reply layers/+lang/dart/packages.el]: Its fine then. I didn't know about that.

**Comment by Compro-Prasad** [comment]
> I'm not sure how it would be possible to offer automatic installation when opening a file tho.

`/layers/auto-layer.el` seems the right place to offer to install.

**Comment by bltavares** [comment]
@Compro-Prasad Cool. I've change it there as well.

**Comment by dennismayr** [comment]
This looks promising!

Let's hope it gets merged soon

**Comment by GLodi** [comment]
Is there any new info regarding the merge? I was hoping to start developing in Flutter with this.

**Comment by duianto** [comment]
Thank you for contributing to Spacemacs!
The PR has been cherry-picked into the `develop` branch, you can safely delete your branch.

**Comment by dennismayr** [comment]
I made an _xpm_ icon for `dart-mode` to include in `mode-icons`.

In `mode-icons-el`, there's a section to define the mode's buffer name, the icon name, and the icon type; example:
```elisp
(defcustom mode-icons
  `(("\\`CSS\\'" "css" xpm)
    ("\\`Coffee\\'" "coffee" xpm-bw)
```
I added a definition:
```elisp
    ("\\`Dart\\'" "dart" xpm) ;; where I added dart.xpm in the `icons` folder
 ```
It's currently not working because perhaps `Dart` isn't the mode's buffer name (e.g. `rjsx-mode` names its buffer `RJSX` instead of `React`).

If you have the answer, I'll open a PR for the icon :smile:

**Comment by dennismayr** [comment]
Question: does this layer include the functionality provided by [flutter.el](https://github.com/amake/flutter.el)? That is, `flutter-run` to dispatch the current project to Gradle, hot-reloading and such, without ever leaving Emacs :)

I've manually added `flutter` to `dotspacemacs-additional-packages` and configured it with `(use-package flutter
    :after dart-mode)` and ` '(flutter-sdk-path "~/flutter/" t)` just in case, but I don't know if my config is being redundant.

**Comment by GLodi** [comment]
They are two separate packages. Your configuration should be good.

Ideally you could make a "flutter" layer that includes both, but even adding flutter to additional-packages it's ok.

**Comment by aislanmaia** [comment]
@GLodi for me spacemacs is outputting the message `No LSP server for dart-mode` when I open up a dart file from a flutter project.

**Comment by duianto** [comment]
@aislanmaia 

When I got that message, it said to check the `*lsp-log*` buffer, and it showed:
>Command "c:/Users/username/Pub/Cache/bin/dart_language_server" is not present on the path.
Command "c:/Users/username/flutter/bin/cache/dart-sdk/bin/dart c:/Users/username/flutter/bin/cache/dart-sdk/bin/snapshots/analysis_server.dart.snapshot --lsp" is not present on the path.
Command "c:/Users/username/Pub/Cache/bin/dart_language_server" is not present on the path.
Command "c:/Users/username/flutter/bin/cache/dart-sdk/bin/dart c:/Users/username/flutter/bin/cache/dart-sdk/bin/snapshots/analysis_server.dart.snapshot --lsp" is not present on the path.

LSP might be assuming that flutter, dart-sdk and the dart language server are located in the home directory. Or it might be because I've added a `HOME` directory environment variable that's pointing to my username folder.

What fixed it was to specify the path to the `dart_language_server` and `dart-sdk` with the LSP dart variables:
```elisp
     (lsp :variables
          lsp-clients-dart-server-command "~/AppData/Roaming/Pub/Cache/bin/dart_language_server"
          lsp-dart-analysis-sdk-dir "c:/src/flutter/bin/cache/dart-sdk/")
```

They should probably be listed in the dart layers readme:
https://github.com/syl20bnr/spacemacs/tree/develop/layers/%2Blang/dart
I'm not sure how it should be described or what the default paths should say.

If anyone has an idea, feel free to open a pull request or suggest something and it can be added.

**Comment by aislanmaia** [comment]
@duianto thanks my friend. With your instructions the lsp-dart is working fine. The problem now seems to be that it is assuming `flutter` path in the wrong dir, actually, an unexistent dir. 
When I'm trying to run `flutter run` from `SPC SPC` it returns `Searching for program: No such file or directory, ~/Development/flutterbin/flutter`. This path it is trying to look at doesn't event exist. My flutter path is `~/Development/flutter/flutter/bin`.

**Comment by duianto** [comment]
Make sure that you have the path to flutter/bin in your OS PATH variable, for me (in Windows) it's in:
`C:\src\flutter\bin`

It might also have to be defined on the `PATH=` line in `.spacemacs.env`, which you can open with 
`SPC f e e`. I might have added the flutter/bin path there manually.

When I press `, x x` (which calls `flutter-run-or-hot-reload`) from for example the file:
`c:/src/flutter/examples/hello_world/test/hello_test.dart`

Then a `*Flutter*` buffer opens with the message:
>Unable to locate a development device; please run 'flutter doctor' for information about installing additional components.
>
>Process Flutter exited abnormally with code 1

So the `flutter` command seems to be detected.

I'm guessing that an android emulator might be included with android studio, which was suggested when I ran `flutter doctor`.

**Pull request merged**